File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
    class Course {
    public:
        int status;
        vector<Course*> prereqs;

        Course() {
            status = 0;
            prereqs = vector<Course*>();
        }
    };

    bool canFinish(Course *course) {
        course->status = 1;
        for (auto prereq: course->prereqs) {
            if (   0 == prereq->status && !canFinish(prereq)
                || 1 == prereq->status) {
                return false;
            }
        }

        course->status = 2;
        return true;
    }

public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        /*
        sort(prerequisites.begin(), prerequisites.end());

        bool erased = true;
        while (erased) {
            erased = false;
            for (int i = 0; prerequisites.size() > i; i++) {
                int required = prerequisites[i][1];
                int j = 0, k = prerequisites.size();
                while (j < k) {
                    int m = (j + k) / 2;
                    if (required == prerequisites[m][0]) {
                        break;
                    }

                    if (required < prerequisites[m][0]) {
                        k = m;
                    } else {
                        j = m + 1;
                    }
                }

                if (j >= k) {
                    prerequisites.erase(prerequisites.begin() + i);
                    erased = true;
                }
            }
        }

        return 0 == prerequisites.size();
        */

        Course *courses = new Course[numCourses];

        for (auto prerequisite: prerequisites) {
            courses[prerequisite[0]].prereqs.push_back(&courses[prerequisite[1]]);
        }

        for (int i = 0; numCourses > i; i++) {
            if (0 == courses[i].status && !canFinish(&courses[i])) {
                return false;
            }
        }

        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def findPrerequisites(self, course: int) -> List[int]:
        res = []
        begin = 0
        end = len(self.prerequisites)
        while begin < end:
            middle = (begin + end) // 2
            if course < self.prerequisites[middle][0]:
                end = middle
                continue
            
            if course > self.prerequisites[middle][0]:
                begin = middle + 1
                continue

            while 0 <= middle and course == self.prerequisites[middle][0]:
                middle -= 1

            middle += 1

            while len(self.prerequisites) > middle and course == self.prerequisites[middle][0]:
                res.append(self.prerequisites[middle][1])
                middle += 1

            break

        return res

    def check(self, course) -> bool:
        if 2 == self.checked[course]:
            return True

        if 1 == self.checked[course]:
            return False

        self.checked[course] = 1
        prereqs = self.findPrerequisites(course)
        for prereq in prereqs:
            if not self.check(prereq):
                return False

        self.checked[course] = 2
        return True

    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        self.prerequisites = sorted(prerequisites)
        self.checked = [0] * numCourses
        
        for i in range(numCourses):
            if not self.check(i):
                return False

        return True
        